### PR TITLE
sql: fix the handling of tuples in distsql

### LINKED ISSUE #25522: sql: Pretty Printing of Tuples does not match postgres
We do not pretty print tuples the same way that postgres does.

Postgres:
```sql
SELECT (1, 2, 'hello', NULL, NULL), (true, NULL, (false, 6.6, false));
      row      |       row
---------------+------------------
 (1,2,hello,,) | (t,,"(f,6.6,f)")
```

Cockroach:
```sql
SELECT (1, 2, 'hello', NULL, NULL), (true, NULL, (false, 6.6, false));
+-----------------------------+--------------------------------+
| (1, 2, 'hello', NULL, NULL) |   (true, NULL, (false, 6.6,    |
|                             |            false))             |
+-----------------------------+--------------------------------+
| (1,2,'hello',,)             | (true,,(false, 6.6, false))    |
+-----------------------------+--------------------------------+
```

### LINKED ISSUE #26624: sql: tuples not properly preserved during serialization
Found while investigating #26621.
The expression `ROW(x)` gets serialized after normalization (format with `FmtParsable`) to `(x)` which has a completely different meaning. This causes the following error when distribution is enabled, but not with local execution:

``` sql
> select row(1) in (row(2), row(1), row(v)) from kv;
pq: (1:::INT) IN ((2:::INT), (1:::INT), ROW("$0")): unsupported comparison operator: (1:::INT) IN ((2:::INT), (1:::INT), ROW("$0")): expected ROW("$0") to be of type int, found type tuple{int}
```

Also all the labeled tuples lose their labels during serialization. This likely breaks distribution of queries with labeled tuples.

### PR DESCRIPTION
**Pull request by knz**
Fixes #26624.
Supersedes #28100.

This patch fixes the bug where 0-valued and 1-valued tuples were not
properly serialized in distsql processor/flow specs, causing these
special tuples to not be properly supported for distributed execution.

It also fixes the bug where labeled DTuples were not serialized with
their tuples, causing labeled tuples to be broken with distributed
execution.

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
Ok I was a bit too trigger happy pushing this PR to everyone. There are really multiple changes/fixes in there. I will split them, add some missing tests and submit the changes separately.

**Review by rjnn — DISMISSED** [review]
:lgtm_strong: on the basic approach, thank you for fixing the underlying problem. However, I agree that this PR should be split up, as there are multiple failing tests on different issues, (which @knz and I discussed offline). All the various little fixes are good, but should probably not be packaged in a single PR.

> Reviewed 31 of 31 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-:-LIpyzm27AAFEy9jqJoT:b-wzuucg)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/expr.go, line 741 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIpxqhF2A_1vyl9l1Ek:-LIpxqhG54DYOB49sLnl:bgpk4pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/65a18f015d870e5c64b32cb44a6ac66c9efe22d2/pkg/sql/sem/tree/expr.go#L741)):*
> ```Go
> 	// col_name.go. It is also used when pretty-printing for human
> 	// consumption, see pretty.go.
> 	Row bool
> ```

Is this bool needed any more?

File: pkg/sql/insert.go [hunk 1]
```diff
@@ -746,35 +746,34 @@ func fillDefaults(
...
 			}
 		}
 
 		// The values for the row may be shorter than the number of columns being
 		// inserted into. Populate default expressions for those columns.
-		for i := len(tuple.Exprs); i < len(cols); i++ {
+		for i := len(tuple); i < len(cols); i++ {
 			copyTuple()
-			tuple.Exprs = append(tuple.Exprs, defaultExpr(len(tuple.Exprs)))
+			ret.Tuples[tIdx] = append(ret.Tuples[tIdx], defaultExpr(len(tuple)))
```
Comment (knz) [inline pkg/sql/insert.go]: @arjunravinarayan to follow-up on our conversation, this is the line where I introduced a bug.

The code should read instead:

```go
tuple = append(tuple, defaultExpr(len(tuple)))
ret.Tuples[tIdx] = tuple
```

**Comment by knz** [comment]
OK PR has been split and simplified. RFAL!

**Comment by knz** [comment]
(the first 3 commits are from different PRs but are prerequisites to this. I will rebase when those merge.)

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28143)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/expr.go, line 741 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIpxqhF2A_1vyl9l1Ek:-LItvZIa1r4ePvFqCqzH:bvogb1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/65a18f015d870e5c64b32cb44a6ac66c9efe22d2/pkg/sql/sem/tree/expr.go#L741)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Is this bool needed any more?
</blockquote></details>

Yes it is needed to derive a column name automatically like pg in e.g. `select row(123)`.

**Review by jordanlewis — CHANGES_REQUESTED** [review]
Generally LGTM though I'm concerned that some of the CLI outputs don't match postgres. See comments

> Reviewed 1 of 31 files at r1, 34 of 34 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-:-LIuy_b-4hqZwqut3oFG:b3uq4kv)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 418 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuvwZPBkv_1M-gooiN:-LIuvwZPBkv_1M-gooiO:byvt6ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/srfs#L418)):*
> ```
> ("c",1)
> ("b",2)
> ("a",3)
> ```

I think this probably is fine, but FYI this is not what postgres outputs.

```
jordan=# SELECT information_schema._pg_expandarray(ARRAY['c', 'b', 'a']);
 _pg_expandarray
-----------------
 (c,1)
 (b,2)
 (a,3)
(3 rows)
```
___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 545 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuw5Yv8lReatz2eYDI:-LIuw5Yv8lReatz2eYDJ:b-jz0bwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/srfs#L545)):*
> ```
> ("(3,1)",123)
> ("(2,2)",123)
> ("(1,3)",123)
> ```


```
jordan=# SELECT ((i.keys).*, 123) FROM (SELECT information_schema._pg_expandarray(ARRAY[3,2,1]) AS keys) AS i;
    row
-----------
 (3,1,123)
 (2,2,123)
 (1,3,123)
```
___
*[pkg/sql/logictest/testdata/logic_test/tuple, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuwB9c-X0oybEuzI5h:-LIuwB9c-X0oybEuzI5i:b-ug7vt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/tuple#L10)):*
> ```
> ----
> t                u
> (1,2,"hello",,)  (t,,"(f,6.6,f)")
> ```

```
jordan=# SELECT (1, 2, 'hello', NULL, NULL) AS t, (true, NULL, (false, 6.6, false)) AS u;
       t       |        u
---------------+------------------
 (1,2,hello,,) | (t,,"(f,6.6,f)")
```
___
*[pkg/sql/sem/tree/expr.go, line 763 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuyJnc7EGUHZ1-CeCB:-LIuyJnc7EGUHZ1-CeCC:b-rlef1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/56db7a053b041fcbd612597402ace458fb2b5beb/pkg/sql/sem/tree/expr.go#L763)):*
> ```Go
> 	ctx.WriteByte('(')
> 	ctx.FormatNode(&node.Exprs)
> 	if len(node.Exprs) == 1 {
> ```

nit: The equivalent in DTuple has a comment explaining this - if you're going to rebase this again plz add that comment here too.
___
*[pkg/sql/sem/tree/format.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuwPJW3eF_3TcioaSz:-LIuwPJW3eF_3TcioaT-:b-ceri52) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/sem/tree/format.go#L87)):*
> ```Go
> 	// If set, strings will be formatted for being contents of ARRAYs.
> 	// Used internally in combination with FmtPgwireText defined below.
> 	pgwireFormat
> ```

Nice idea.

But this flag does more than the comment says, now. It also prints NULL as the empty string, right?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28143)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 418 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuvwZPBkv_1M-gooiN:-LIvOkRK9KA7k5TBjHyL:bzyr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/srfs#L418)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think this probably is fine, but FYI this is not what postgres outputs.

```
jordan=# SELECT information_schema._pg_expandarray(ARRAY['c', 'b', 'a']);
 _pg_expandarray
-----------------
 (c,1)
 (b,2)
 (a,3)
(3 rows)
```
</blockquote></details>

Discussion for #28151.
My comment there: "it's ugly but it is correct".
___
*[pkg/sql/logictest/testdata/logic_test/srfs, line 545 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuw5Yv8lReatz2eYDI:-LIvOuDM7Af7_z8tUyBt:b-x7sqvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/srfs#L545)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>


```
jordan=# SELECT ((i.keys).*, 123) FROM (SELECT information_schema._pg_expandarray(ARRAY[3,2,1]) AS keys) AS i;
    row
-----------
 (3,1,123)
 (2,2,123)
 (1,3,123)
```
</blockquote></details>

That's a separate bug, well spotted: #28203
___
*[pkg/sql/logictest/testdata/logic_test/tuple, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuwB9c-X0oybEuzI5h:-LIvPonN6tgn3foXW9qd:bbhgbfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/logictest/testdata/logic_test/tuple#L10)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

```
jordan=# SELECT (1, 2, 'hello', NULL, NULL) AS t, (true, NULL, (false, 6.6, false)) AS u;
       t       |        u
---------------+------------------
 (1,2,hello,,) | (t,,"(f,6.6,f)")
```
</blockquote></details>

Discussion for #28151. Same comment as above.
___
*[pkg/sql/sem/tree/expr.go, line 763 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuyJnc7EGUHZ1-CeCB:-LIvQidA98_LcVcZiAcJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56db7a053b041fcbd612597402ace458fb2b5beb/pkg/sql/sem/tree/expr.go#L763)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: The equivalent in DTuple has a comment explaining this - if you're going to rebase this again plz add that comment here too.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/format.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-LIuwPJW3eF_3TcioaSz:-LIvQmAq-E2jyTTctdLW:b-z4ve90) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e592b3e18cf9da631d23388178c34db7142e68e/pkg/sql/sem/tree/format.go#L87)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Nice idea.

But this flag does more than the comment says, now. It also prints NULL as the empty string, right?
</blockquote></details>

Changed the comment. Thanks for noticing.

**Review by jordanlewis — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 27 files at r4, 25 of 25 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28143#-:-LIvaMcR2a5OmI_UnGjc:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

**Comment by knz** [comment]
Thanks all

bors r+

**Pull request merged**